Expand on migration story to indexed-traversable (#961)
ekmett/lens: CHANGELOG.markdown

  Beware: the `optics-core` package (versions <0.4) defines similar classes,
  and will also migrate to use `indexed-traversable` classes. Therefore, you
  might get duplicate instance errors if your package defines both.

>   If you define your own `FunctorWithIndex` etc instances,

Reviewer: ```suggestion
  If you define your own `FunctorWithIndex` etc. instances,
```

>   we recommend to depend directly on the `indexed-traversable` package.

Reviewer: ```suggestion
  we recommend that you depend directly on the `indexed-traversable` package.
```

  If you want to continue support `lens-4` users, you may write

  ```haskell
  -- from indexed-traversable
  import Data.Functor.WithIndex

  -- from lens
  import qualified Control.Lens as L

  -- your (indexed) container
  data MySeq a = ...

  -- indexed-traversable instance
  instance FunctorWithIndex     Int MySeq where imap = ...
  instance FoldableWithIndex    Int MySeq where ifoldMap = ...
  instance TraversableWithIndex Int MySeq where itraverse = ...

  -- lens <5 instance, note the !
  #if !MIN_VERSION_lens(5,0,0)
  instance L.FunctorWithIndex     Int MySeq where imap = imap
  instance L.FoldableWithIndex    Int MySeq where ifoldMap = ifoldMap
  instance L.TraversableWithIndex Int MySeq where itraverse = itraverse
  #endif
  ```

  In other words, always provide `indexed-traversable` instances,

>   and additionally instances for `lens-4.x` type classes,

Reviewer: Nitpick: you refer to `lens-4.x` here, but `lens-4` above. I'd recommend picking a single convention and using it throughout the changelog.

>   and additionally instances for `lens-4.x` type classes,

Reviewer: Perhaps replace "additionally" with "optionally", since the `L.FunctorWithIndex` family of instances aren't strictly required.

Author: True.

Author: Ah, not sure. That paragraph is in one sense continuation of *If you want to continue support `lens-4` users, you may write*, but OTOH also a summary of whole entry.

In first interpreation it is *additionally*, in general it is *optionally*.

Author: I rewrote this paragraph. Is it any better?

  Beware: the `optics-core` package (versions <0.4) defines similar classes,
  and will also migrate to use `indexed-traversable` classes. Therefore, you
  might get duplicate instance errors if your package defines both.

  If you define your own `FunctorWithIndex` etc. instances,
  we recommend that you depend directly on the `indexed-traversable` package.
  If you want to continue support `lens-4` users, you may write

  ```haskell
  -- from indexed-traversable
  import Data.Functor.WithIndex

  -- from lens
  import qualified Control.Lens as L

  -- your (indexed) container
  data MySeq a = ...

  -- indexed-traversable instance
  instance FunctorWithIndex     Int MySeq where imap = ...
  instance FoldableWithIndex    Int MySeq where ifoldMap = ...
  instance TraversableWithIndex Int MySeq where itraverse = ...

  -- lens <5 instance, note the !
  #if !MIN_VERSION_lens(5,0,0)
  instance L.FunctorWithIndex     Int MySeq where imap = imap
  instance L.FoldableWithIndex    Int MySeq where ifoldMap = ifoldMap
  instance L.TraversableWithIndex Int MySeq where itraverse = itraverse
  #endif
  ```

  In other words, always provide `indexed-traversable` instances.

>   If your package depends on `lens`, and allows `lens-4`,

Reviewer: ```suggestion
  If your package depends on `lens` and allows `lens-4`,
```

  you should additionally provide instances for `lens-4` type classes

>   which can be piggyback on `indexed-traversable` instances.

Reviewer: ```suggestion
  that can reuse the `indexed-traversable` instances.
```
